## Session 3: Grouping, Aggregations & Feature Engineering

**Date:** 18/05/2026  
**Objective:** Apply grouping and aggregation techniques to uncover patterns in gaming communities. Create new features (engagement ratio, tier scores, age groups, etc.) to enrich the dataset for deeper analysis. Prepare the data for visualisation and statistical testing.

---

### 3.1 Progress Overview

| Part | Focus | Status |
|------|-------|--------|
| **Part A** | Basic groupby (single column) | ✅ Completed |
| **Part B** | Multiple aggregations & custom functions | ✅ Completed |
| **Part C** | Multi‑level grouping (two or more columns) | ✅ Completed |
| **Part D** | Ranking & filtering within groups | ✅ Completed |
| **Part E** | Feature engineering (new columns) | ✅ Completed |
| **Part F** | Advanced aggregations & pivot tables | ✅ Completed |

---

### 3.2 Completed Tasks (Parts A–F)

#### Part A – Basic GroupBy
- Average `member_count` per genre – **Fighting** is largest (5,596).
- Total tournaments per `community_tier` – **Semi‑Pro** hosts most (525).
- Median active players per `primary_platform` – **Twitter/X** highest (311).
- Min/max average member age per region – **EU** has youngest (14.0).
- Language counts – **English** most common (141).

#### Part B – Multiple Aggregations & Custom Functions
- `.agg()` on genre: mean members, sum tournaments, median rank.
- Top 5 countries by total members – **United States** leads.
- Standard deviation of member counts per tier – **Professional** most varied.
- Distinct primary games per genre – **FPS** most diverse (5 games).
- Discord/Reddit ratio per region – **OC** (Oceania) relies most on Discord.

#### Part C – Multi‑Level Grouping
- Community counts per `genre` + `community_tier` – **FPS Amateur** most common (83).
- Average active players per `region` + `primary_platform` – **AS + Twitch** highest (3,696).
- Total member count per `year_founded` + `genre` – **2019 FPS** peak (158,947).
- Worst rank per `language` + `community_tier` – **English Amateur** has highest (worst) rank.

#### Part D – Ranking & Filtering within Groups
- Largest community per genre – **Reaper Crew (FPS)** largest (68,819).
- Top active community per region – **Pulse Nation (AS)** most active (23,727).
- Genres removed by tournament filter (<50 total) – **MMORPG, Strategy, Fighting**.
- Second largest community per tier – varies from `Dusk Raiders` (Amateur) to `Pulse Nation` (Elite).
- Above region average percentage – only **18.8%** of communities exceed their region's mean.

#### Part E – Feature Engineering (New Columns Created)
- `engagement_ratio` – active players / member count (highest in Battle Royale, 0.45).
- `tier_score` – numeric mapping of community tier (Asia highest, 1.63).
- `age_group` – categorised from avg_member_age (FPS dominates all age groups).
- `community_age` – years since founding (oldest: 18 years in EU).
- `has_coaching` – boolean (100% of Elite/Pro communities have coaching).
- `has_multiple_platforms` – Discord + Reddit presence (51.5% of communities).
- `log_member_count` – log transformation for skewness.
- `tournament_intensity` – tournaments per 1,000 members (mean 8.01).
- `top_player_tier` – quantile‑based rank tiers (High/Medium/Low).
- `total_social_presence` – weighted sum of platforms + content creators.

#### Part F – Advanced Aggregations & Pivot Tables
- **F1:** Pivot table – mean `member_count` by `genre` × `community_tier` (filled NaN with 0).  
- **F2:** Pivot table – count of communities by `region` × `primary_platform`.  
- **F3:** Crosstab – `genre` vs `language` (normalised by row).  
- **F4:** Cumulative sum of `member_count` per genre ordered by `year_founded`.  
- **F5:** Enhanced DataFrame saved as `gaming_communities_enhanced.csv` (index reset, clean ordering).

---

### 3.3 Deliverables by End of Session

- [x] Parts A–D completed with code and insights.
- [x] Part E completed (10 new features).
- [x] Part F completed (pivot tables, crosstab, cumulative sum, enhanced CSV).
- [x] Enhanced dataset saved as `gaming_communities_enhanced.csv`.
- [x] Summary of key findings (e.g., which genre has highest average engagement, which tier is most active).
- [x] At least one pivot table or crosstab with meaningful insight.

---

### 3.4 Next Steps (Preview)

**Session 4:** Visualisation & Statistical Analysis  
- Create static (Matplotlib/Seaborn) and interactive (Plotly) charts.
- Perform hypothesis tests (t‑tests, ANOVA, chi‑square) on groups.
- Build a small dashboard to present key metrics.

---

### 3.5 Tools & Libraries

- Python 3.10+
- pandas, numpy
- (matplotlib, seaborn, plotly for Session 4)

---

*This notebook completes the grouping, aggregation, and feature engineering phase. The enhanced dataset is ready for visualisation and statistical testing in Session 4.*

In [129]:
#importing essential libraries
import pandas as pd
import numpy as np

In [130]:
#loading the dataset
gaming_data = pd.read_csv('gaming_communities_data_cleaned.csv')

In [131]:
genre = gaming_data.groupby(['genre'])
print('Genre with the highest average community size:')
genre['member_count'].mean().nlargest(1)

Genre with the highest average community size:


genre
Fighting    5596.4375
Name: member_count, dtype: float64

Genre with the highest average community size is Fighting (5596.437), followed by MOBA (3551.31), FPS (3242.57) and so on 

In [132]:
community_tier = gaming_data.groupby(['community_tier'])
community_tier['tournaments_hosted'].sum()

community_tier
Amateur          20
Elite            21
Professional    123
Semi-Pro        525
Name: tournaments_hosted, dtype: int64

Semi‑Pro communities host the most tournaments (525), followed by Professional (123). Amateur and Elite host very few.

In [133]:
prim_platform = gaming_data.groupby(['primary_platform'])
print('Platform with the highest average active players:')
prim_platform['active_players'].median().nlargest(1)

Platform with the highest average active players:


primary_platform
Twitter/X    311.0
Name: active_players, dtype: float64

Platform with the highest average of active players is Twitter/X (311.0), followed by Twitch (303.0), YouTube (254.5) and so on

In [134]:
region = gaming_data.groupby(['region'])
region['avg_member_age'].agg(['min', 'max'])

,min,max
region,,
AF,19.1,31.8
AS,15.7,29.9
EU,14.0,37.9
OC,17.1,34.6
SA,14.7,32.6
Unknown,14.0,38.3


The region with the youngest member average age is EU (europe) with minimum avg age of 14.0 and maximum avg age of 37.9

In [135]:
language = gaming_data.groupby(['language'])
language['community_name'].count().sort_values(ascending=False)

language
English       141
Spanish        62
Portuguese     32
Chinese        28
Korean         24
Russian        19
Arabic         14
French         13
German         13
Japanese       11
Name: community_name, dtype: int64

The top three most common languages used by communities are English (141), Spanish (62) and Portuguese (32)

In [136]:
gaming_data.groupby(['genre']).agg({
    'member_count': 'mean',
    'tournaments_hosted': 'sum',
    'top_player_rank': 'median'
})

,member_count,tournaments_hosted,top_player_rank
genre,,,
Battle Royale,3241.642857,103,3505.5
FPS,3242.579618,361,5025.0
Fighting,5596.437500,20,4581.0
MMORPG,2427.523810,4,4276.0
MOBA,3551.312500,150,3428.0
Sports,3059.240000,51,4693.0
Strategy,346.750000,0,3634.5


In [137]:
gaming_data.groupby(['country_of_origin'])['member_count'].sum().sort_values(ascending=False).iloc[:5]

country_of_origin
United States    202982
Taiwan           174737
Mexico           146408
China            104804
South Africa      74034
Name: member_count, dtype: int64

In [138]:
gaming_data.groupby(['community_tier'])['member_count'].std().sort_values(ascending=False)

community_tier
Professional    18501.523943
Semi-Pro         3476.735288
Amateur           498.502645
Elite               0.000000
Name: member_count, dtype: float64

In [139]:
gaming_data.groupby(['genre'])['primary_game'].nunique().sort_values(ascending=False)

genre
FPS              5
Battle Royale    3
MOBA             3
Fighting         2
MMORPG           2
Sports           1
Strategy         1
Name: primary_game, dtype: int64

In [140]:
gaming_data['discord_to_reddit_ratio'] = gaming_data['discord_server_size'] / gaming_data['reddit_community_size']

gaming_data['discord_to_reddit_ratio'] = gaming_data['discord_to_reddit_ratio'].replace([np.inf, -np.inf], 0).fillna(0)

gaming_data.groupby(['region'])['discord_to_reddit_ratio'].mean()

region
AF         1.701499
AS         1.833267
EU         2.299356
OC         3.243439
SA         2.069069
Unknown    2.317666
Name: discord_to_reddit_ratio, dtype: float64

In [141]:
gaming_data.groupby(['genre', 'community_tier']).size().reset_index(name='count')

,genre,community_tier,count
0,Battle Royale,Amateur,40
1,Battle Royale,Professional,6
2,Battle Royale,Semi-Pro,24
3,FPS,Amateur,83
4,FPS,Elite,3
5,FPS,Professional,12
6,FPS,Semi-Pro,59
7,Fighting,Amateur,8
8,Fighting,Professional,1
9,Fighting,Semi-Pro,7


In [142]:
gaming_data.groupby(['region', 'primary_platform'])['active_players'].mean().reset_index(name='Average').sort_values(by='Average', ascending=False)

,region,primary_platform,Average
6,AS,Twitch,3695.571429
8,AS,YouTube,2290.666667
26,Unknown,Twitch,1779.153846
0,AF,Discord,1592.090909
25,Unknown,Reddit,1429.111111
10,EU,Reddit,1341.300000
14,OC,Discord,1290.571429
4,AS,Discord,1215.363636
3,AF,YouTube,1133.750000
2,AF,Twitch,1059.000000


In [143]:
gaming_data.groupby(['year_founded', 'genre'])['member_count'].sum().reset_index(name='Total')

,year_founded,genre,Total
0,2008,MMORPG,55
1,2010,MMORPG,3074
2,2012,MOBA,100862
3,2012,Sports,282
4,2013,MMORPG,1584
5,2013,MOBA,2021
6,2014,MOBA,6510
7,2015,Battle Royale,18092
8,2015,FPS,44320
9,2015,MMORPG,367


In [144]:
gaming_data.groupby(['language', 'community_tier'])['top_player_rank'].max().reset_index(name='Worst_rank').nlargest(n=1, columns='Worst_rank')

,language,community_tier,Worst_rank
7,English,Amateur,9843


In [145]:
idx = gaming_data.groupby(['genre'])['member_count'].idxmax()
gaming_data.loc[idx, ['genre', 'community_name', 'member_count']]

,genre,community_name,member_count
103,Battle Royale,Blizzard Hawks,52801
28,FPS,Reaper Crew,68819
346,Fighting,Raven Guard,58240
72,MMORPG,Warpath,15277
291,MOBA,Spike Rush,68819
146,Sports,Vortex,47742
321,Strategy,Rune Masters,657


- In the **FPS** genre, "Reaper Crew" is the largest community (68,819 members).  
- The **Strategy** genre has the smallest largest community (657 members), reflecting its niche audience.  
- Fighting and Battle Royale also have very large communities (>50k members).

In [146]:
gaming_data['region_activity_rank'] = gaming_data.groupby(['region'])['active_players'].rank(ascending=False, method='dense')
top_per_region = gaming_data[gaming_data['region_activity_rank'] == 1][['region', 'community_name', 'active_players']]
top_per_region

,region,community_name,active_players
105,AS,Pulse Nation,23727
146,OC,Vortex,10874
188,EU,Iron Phoenix,11475
245,Unknown,Sunfire,14589
291,AF,Spike Rush,13865
349,SA,Comet Strike,4811


- **Asia (AS)** has the most active community ("Pulse Nation" with 23,727 active players).  
- **Unknown** region also shows high activity ("Sunfire" with 14,589).  
- **South America (SA)** has the lowest top activity (4,811), suggesting smaller active player bases.

In [147]:
filtered = gaming_data.groupby(['genre']).filter(lambda x: x['tournaments_hosted'].sum() > 50)
print(f"Removed Genres: {set(gaming_data['genre']) - set(filtered['genre'])}")

Removed Genres: {'MMORPG', 'Strategy', 'Fighting'}


- Genres removed (total tournaments ≤ 50): **MMORPG, Strategy, Fighting**.  
- These genres host very few tournaments overall, so they may be less competitive or have different community structures.

In [148]:
sorted_data = gaming_data.sort_values(by=['community_tier', 'member_count'], ascending=[True, False])

second_per_tier = sorted_data.groupby('community_tier').nth(1).reset_index()

second_per_tier[['community_tier', 'community_name', 'member_count']]

,community_tier,community_name,member_count
0,Amateur,Dusk Raiders,1879
1,Elite,Pulse Nation,68819
2,Professional,Blizzard Hawks,52801
3,Semi-Pro,War Hammer,13712


- In the **Amateur** tier, the second largest community is "Dusk Raiders" (1,879 members).  
- For **Elite**, it's "Pulse Nation" (68,819) – note that Elite has only 3 communities, so second largest is actually the second best.  
- **Semi-Pro** second largest is "War Hammer" (13,712 members).


In [149]:
region_avg = gaming_data.groupby(['region'])['member_count'].transform('mean')
gaming_data['above_region_avg'] = gaming_data['member_count'] > region_avg
print(f"Percentage above region average: {gaming_data['above_region_avg'].mean()*100:.1f}%")

Percentage above region average: 18.8%


- Only **18.8%** of communities have member count above their region's average.  
- This means most communities are smaller than the regional average – the average is pulled up by a few very large communities (outliers).  
- Regions with high inequality (e.g., one giant community) will have a low percentage.

In [150]:
gaming_data['engagement_ratio'] = gaming_data['active_players'] / gaming_data['member_count']
gaming_data.groupby(['genre'])['engagement_ratio'].mean()

genre
Battle Royale    0.454207
FPS              0.446840
Fighting         0.393940
MMORPG           0.452380
MOBA             0.405555
Sports           0.445047
Strategy         0.374489
Name: engagement_ratio, dtype: float64

- **Battle Royale** and **MMORPG** have the highest engagement ratios (~0.45), meaning nearly half of their members are active players.  
- **Strategy** has the lowest engagement (0.37), possibly due to smaller, more casual communities.

In [151]:
gaming_data['tier_score'] = gaming_data['community_tier'].map({
    'Amateur': 1, 
    'Semi-Pro': 2, 
    'Professional': 3, 
    'Elite': 4
})

gaming_data.groupby(['region'])['tier_score'].mean()

region
AF         1.565217
AS         1.633803
EU         1.536842
OC         1.434783
SA         1.367347
Unknown    1.552083
Name: tier_score, dtype: float64

- **Asia (AS)** has the highest average tier score (1.63), meaning its communities are closest to Semi‑Pro on average.  
- **South America (SA)** has the lowest (1.37), indicating more Amateur-level communities.

In [152]:
bins = [0, 18, 25, 35, 100]
labels = ['Teen', 'Young Adult', 'Adult', 'Senior']

gaming_data['age_group'] = pd.cut(gaming_data['avg_member_age'], bins=bins, labels=labels)
gaming_data.groupby(['genre'])['age_group'].value_counts().unstack()

age_group,Teen,Young Adult,Adult,Senior
genre,,,,
Battle Royale,10,48,12,0
FPS,23,108,26,0
Fighting,2,7,7,0
MMORPG,0,3,17,1
MOBA,1,45,17,1
Sports,3,11,11,0
Strategy,0,1,3,0


- **FPS** is the most popular genre across all age groups, especially among Young Adults (108 communities).  
- **MMORPG** and **MOBA** have more Adult and Senior members compared to other genres.

In [153]:
gaming_data['community_age'] = 2026 - gaming_data['year_founded'] 

idx_oldest = gaming_data.groupby(['region'])['community_age'].idxmax()
idx_youngest = gaming_data.groupby(['region'])['community_age'].idxmin()

oldest = gaming_data.loc[idx_oldest, ['region', 'community_name', 'community_age']].assign(status='Oldest')
youngest = gaming_data.loc[idx_youngest, ['region', 'community_name', 'community_age']].assign(status='Youngest')

pd.concat([oldest, youngest]).sort_values(by=['region', 'status'])

,region,community_name,community_age,status
291,AF,Spike Rush,14,Oldest
186,AF,Gravity Wave,2,Youngest
182,AS,Bright Star,14,Oldest
119,AS,Raging Tempest,2,Youngest
209,EU,Bright Moon,18,Oldest
37,EU,Starfall,2,Youngest
184,OC,Solar Flare,12,Oldest
112,OC,Steel Phoenix,2,Youngest
43,SA,Iron Sharks,16,Oldest
76,SA,Emerald Legion,2,Youngest


- The **oldest** communities (founded 2008–2010) are in **AF**, **AS**, **EU**, and **Unknown** regions.  
- The **youngest** communities (founded 2024) exist in every region, showing continuous growth.  
- **EU** has the widest age range (2 to 18 years).

In [154]:
gaming_data['total_social_presence'] = gaming_data['discord_server_size'] + gaming_data['reddit_community_size'] + (gaming_data['content_creators'] * 1000)
gaming_data.nlargest(5, 'total_social_presence')[['community_name', 'total_social_presence']]

,community_name,total_social_presence
105,Pulse Nation,123596
28,Reaper Crew,101286
245,Sunfire,99763
291,Spike Rush,73211
346,Raven Guard,54913


- **Pulse Nation** has the highest total social presence (123,596), driven by a massive Discord server (95,596) and many content creators (28 × 1000).  
- **Reaper Crew** and **Sunfire** also have >99k scores, indicating they are highly active across platforms.

In [155]:
gaming_data['has_coaching'] = gaming_data['coaching_staff'] > 0
gaming_data.groupby(['community_tier'])['has_coaching'].mean() * 100

community_tier
Amateur           0.00000
Elite           100.00000
Professional    100.00000
Semi-Pro         66.40625
Name: has_coaching, dtype: float64

- **100%** of Elite and Professional communities have coaching staff.  
- **66.4%** of Semi‑Pro communities have coaching, while **0%** of Amateur communities do – coaching is a strong differentiator of competitive tiers.

In [156]:
gaming_data['has_multiple_platforms'] = (gaming_data['discord_server_size'] > 0) & (gaming_data['reddit_community_size'] > 0)
print(f"Percentage with multiple platforms: {gaming_data['has_multiple_platforms'].mean()*100:.1f}%")


Percentage with multiple platforms: 51.5%


- Only **51.5%** of communities maintain both a Discord server and a Reddit presence. Most focus on a single primary platform.

In [157]:
gaming_data['log_member_count'] = np.log1p(gaming_data['member_count'])

In [158]:
gaming_data['tournament_intensity'] = gaming_data['tournaments_hosted'] / (gaming_data['member_count'] / 1000)
print(f"Mean tournament intensity: {gaming_data['tournament_intensity'].mean():.2f}")

Mean tournament intensity: 8.01


In [159]:
gaming_data['top_player_tier'] = pd.qcut(gaming_data['top_player_rank'], q=3, labels=['High', 'Medium', 'Low'])
gaming_data.groupby(['genre'])['top_player_tier'].value_counts().unstack()

top_player_tier,High,Medium,Low
genre,,,
Battle Royale,26,27,17
FPS,47,50,60
Fighting,4,8,4
MMORPG,9,5,7
MOBA,28,14,22
Sports,4,13,8
Strategy,1,2,1


- **FPS** has the most communities in the "Low" rank (worst) tier (60), but also many in "Medium" and "High".  
- **Battle Royale** and **MOBA** are more balanced across tiers.  
- **Fighting** and **Strategy** have few communities overall, with most in "Medium".

In [160]:
pivot = pd.pivot_table(
    gaming_data,
    index='genre',
    columns='community_tier',
    values='member_count',
    aggfunc='mean',
    fill_value=0
)
pivot

community_tier,Amateur,Elite,Professional,Semi-Pro
genre,,,,
Battle Royale,574.725000,0.0,26590.666667,1849.250000
FPS,501.253012,68819.0,8584.333333,2678.169492
Fighting,990.000000,0.0,58240.000000,3340.428571
MMORPG,601.066667,0.0,15277.000000,5337.000000
MOBA,561.916667,68819.0,16909.000000,3069.565217
Sports,405.785714,0.0,47742.000000,2305.800000
Strategy,346.750000,0.0,0.000000,0.000000


In [161]:
pivot = pd.pivot_table(
    gaming_data,
    index='region',
    columns='primary_platform',
    values='community_name',
    aggfunc='count',
    fill_value=0
)
pivot

primary_platform,Discord,Reddit,Twitch,Twitter/X,YouTube
region,,,,,
AF,11,6,2,0,4
AS,44,7,7,7,6
EU,61,10,10,7,7
OC,14,3,3,1,2
SA,22,9,8,3,7
Unknown,61,9,13,9,4


In [162]:
crosstab = pd.crosstab(
    gaming_data['genre'],
    gaming_data['language'],
    normalize='index'
)
crosstab

language,Arabic,Chinese,English,French,German,Japanese,Korean,Portuguese,Russian,Spanish
genre,,,,,,,,,,
Battle Royale,0.028571,0.057143,0.342857,0.028571,0.057143,0.014286,0.185714,0.014286,0.071429,0.200000
FPS,0.025478,0.108280,0.388535,0.044586,0.025478,0.012739,0.044586,0.095541,0.070064,0.184713
Fighting,0.125000,0.000000,0.312500,0.000000,0.000000,0.125000,0.000000,0.062500,0.062500,0.312500
MMORPG,0.047619,0.047619,0.571429,0.000000,0.047619,0.000000,0.047619,0.095238,0.095238,0.047619
MOBA,0.078125,0.078125,0.437500,0.031250,0.031250,0.093750,0.031250,0.109375,0.000000,0.109375
Sports,0.000000,0.040000,0.400000,0.080000,0.080000,0.000000,0.040000,0.200000,0.000000,0.160000
Strategy,0.000000,0.000000,0.250000,0.000000,0.000000,0.000000,0.000000,0.250000,0.000000,0.500000


In [163]:
gaming_data = gaming_data.sort_values(by=['genre', 'year_founded'])
gaming_data['cumulative_member_by_genre'] = gaming_data.groupby(['genre'])['member_count'].cumsum()

In [164]:
if 'index' in gaming_data.columns:
    gaming_data = gaming_data.drop(columns=['index'])

# 2. Sort the dataset back by your community_id so it starts from 1 and goes perfectly in order
gaming_data = gaming_data.sort_values(by='community_id').reset_index(drop=True)

# 3. Save it again cleanly without any extra index columns tracking the old mess
gaming_data.to_csv('gaming_communities_enhanced.csv', index=False)
print("Dataset completely restored! 'index' column removed, and ordered from 1 onwards.")

Dataset completely restored! 'index' column removed, and ordered from 1 onwards.
